unittest/server/src/ivi_id_agent_unittests.cpp

```diff
@@ -0,0 +1,578 @@
+
+#include <gtest/gtest.h>
+#include <string>
+#include "server_api_fake.h"
+#include "ivi_layout_interface_fake.h"
+
+extern "C"
+{
+#include "ivi-id-agent.c"
+}
+
+#define INVALID_ID 0xFFFFFFFF
+static constexpr uint8_t MAX_NUMBER = 2;
+
+class IdAgentTest: public ::testing::Test
+{
+public:
+    void SetUp()
+    {
+        // ASSERT_EQ(initBaseModule(), true);
+        IVI_LAYOUT_FAKE_LIST(RESET_FAKE);
+        SERVER_API_FAKE_LIST(RESET_FAKE);
+        init_controller_content();
+    }
+
+    void TearDown()
+    {
+        deinit_controller_content();
+    }
+
+    void init_controller_content()
+    {
+        m_iviShell.interface = &g_iviLayoutInterfaceFake;
+        m_iviShell.compositor = &m_westonCompositor;
+        custom_wl_list_init(&m_westonCompositor.seat_list);
+        custom_wl_list_init(&m_iviShell.list_surface);
+        custom_wl_list_init(&m_westonCompositor.seat_list);
+        wl_signal_init(&m_iviShell.id_allocation_request_signal);
+        wl_signal_init(&m_iviShell.ivisurface_created_signal);
+        wl_signal_init(&m_iviShell.ivisurface_removed_signal);
+        mp_iviIdAgent = (struct ivi_id_agent*)malloc(sizeof(struct ivi_id_agent));
+        mp_iviIdAgent->compositor = &m_westonCompositor;
+        mp_iviIdAgent->interface = &g_iviLayoutInterfaceFake;
+        mp_iviIdAgent->default_surface_id = 100;
+        mp_iviIdAgent->default_surface_id_max = 200;
+        mp_iviIdAgent->default_behavior_set = 0;
+
+        custom_wl_list_init(&mp_iviIdAgent->app_list);
+        custom_wl_list_init(&mp_iviIdAgent->id_allocation_listener.link);
+        custom_wl_list_init(&mp_iviIdAgent->surface_removed.link);
+        custom_wl_list_init(&mp_iviIdAgent->destroy_listener.link);
+        custom_wl_list_init(&m_westonCompositor.destroy_signal.listener_list);
+
+        for(uint8_t i = 0; i < MAX_NUMBER; i++)
+        {
+            // prepare for desktop apps
+            mp_dbElem[i] = (struct db_elem*)malloc(sizeof(struct db_elem));
+            mp_dbElem[i]->surface_id = 10 + i;
+            mp_dbElem[i]->cfg_app_id = (char*)malloc(5);
+            mp_dbElem[i]->cfg_title = (char*)malloc(10);
+            snprintf(mp_dbElem[i]->cfg_app_id, 5, "%d", i);
+            snprintf(mp_dbElem[i]->cfg_title, 10, "idtest%d", i);
+            custom_wl_list_insert(&mp_iviIdAgent->app_list, &mp_dbElem[i]->link);
+        }
+    }
+
+    void deinit_controller_content()
+    {
+        for(uint8_t i = 0; i < MAX_NUMBER; i++)
+        {
+            if(mp_dbElem[i] != nullptr)
+            {
+                free(mp_dbElem[i]->cfg_app_id);
+                free(mp_dbElem[i]->cfg_title);
+                free(mp_dbElem[i]);
+            }
+        }
+        if(mp_iviIdAgent != nullptr)
+        {
+            free(mp_iviIdAgent);
+        }
+    }
+
+    struct ivishell m_iviShell = {};
+    struct ivi_id_agent *mp_iviIdAgent = nullptr;
+    struct weston_compositor m_westonCompositor = {};
```

Comment:
may consider to have a common function to setup weston_compositor structure for all server test